File: agenkit-scala/src/main/scala/io/agenkit/patterns/SupervisorAgent.scala
package io.agenkit.patterns

import io.agenkit.adapters.LlmClient
import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Decomposes a task and delegates sub-tasks to a pool of worker agents. */
class SupervisorAgent(
  val name: String,
  llm: LlmClient,
  workers: List[Agent]
) extends Agent:
  private val _processedCount = new AtomicLong(0)

  def capabilities: List[String] = List("supervision", "decomposition", "coordination")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    val workerNames = workers.map(_.name).mkString(", ")
    val decomposePrompt = Message.system(
      s"Decompose the task into subtasks, one per worker. Workers: $workerNames. Format: WORKER_NAME: subtask"
    )
    llm.complete(List(decomposePrompt, message)).flatMap { plan =>
      val assignments = parseAssignments(plan.contentString)
      val workerMap   = workers.map(w => w.name.toLowerCase -> w).toMap
      val futures = assignments.flatMap { case (workerName, task) =>
        workerMap.get(workerName.toLowerCase).map(_.process(Message.user(task)))
      }
      Future.sequence(futures).map { results =>
        val combined = results.zipWithIndex
          .map { case (r, i) => s"Worker ${i + 1}: ${r.contentString}" }
          .mkString("\n")
        Message.of("assistant", combined)
      }
    }

  private def parseAssignments(text: String): List[(String, String)] =
    text.split("\n").flatMap { line =>
      line.split(":", 2) match
        case Array(worker, task) => Some(worker.trim -> task.trim)
        case _                   => None
    }.toList

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("workerCount" -> workers.size, "workers" -> workers.map(_.name)),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/patterns/SupervisorAgent.java
package io.agenkit.patterns;

import io.agenkit.adapters.LlmClient;
import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * A supervisor agent that decomposes tasks and delegates to worker agents.
 */
public final class SupervisorAgent implements Agent {

    private final String name;
    private final LlmClient llmClient;
    private final List<Agent> workers;

    public SupervisorAgent(String name, LlmClient llmClient, List<Agent> workers) {
        this.name = name;
        this.llmClient = llmClient;
        this.workers = List.copyOf(workers);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("supervision", "task_decomposition", "delegation");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        StringBuilder workerDesc = new StringBuilder("Available workers:\n");
        for (Agent worker : workers) {
            workerDesc.append("- ").append(worker.getName())
                    .append(": ").append(worker.getCapabilities()).append("\n");
        }

        List<Message> planMessages = List.of(
                Message.of("system", workerDesc + "\nDecompose the task. Format: WORKER_NAME: subtask"),
                message);

        return llmClient.complete(planMessages).thenCompose(plan -> {
            List<CompletableFuture<String>> tasks = new ArrayList<>();
            for (String line : plan.contentString().split("\n")) {
                for (Agent worker : workers) {
                    if (line.startsWith(worker.getName() + ":")) {
                        String subtask = line.substring(worker.getName().length() + 1).trim();
                        tasks.add(worker.process(Message.of("user", subtask))
                                .thenApply(r -> worker.getName() + ": " + r.contentString()));
                        break;
                    }
                }
            }

            if (tasks.isEmpty()) {
                // Fall back to first worker
                Agent first = workers.isEmpty() ? null : workers.get(0);
                if (first == null) {
                    return CompletableFuture.completedFuture(
                            Message.of("assistant", "No workers available"));
                }
                return first.process(message)
                        .thenApply(r -> r.withMetadata("supervisor", name));
            }

            return CompletableFuture.allOf(tasks.toArray(new CompletableFuture[0]))
                    .thenApply(v -> {
                        List<String> results = tasks.stream()
                                .map(CompletableFuture::join)
                                .toList();
                        return Message.of("assistant", String.join("\n", results))
                                .withMetadata("supervisor", name)
                                .withMetadata("workers_used", tasks.size());
                    });
        });
    }

    @Override
    public IntrospectionResult introspect() {
        List<String> workerNames = workers.stream().map(Agent::getName).toList();
        Map<String, Object> state = new HashMap<>();
        state.put("workerCount", workers.size());
        return new IntrospectionResult(name, getCapabilities(), null, state, workerNames);
    }
}
